toothpick-runtime/src/main/java/toothpick/ScopeImpl.java

```diff
@@ -390,47 +389,66 @@ private void installModule(Module module) {
    * and {@link #installUnBoundProvider(Class, String, InternalProviderImpl)}
    * are a facade of this method and make the calls more clear.
    */
-  private <T> void installInternalProvider(Class<T> clazz, String bindingName, InternalProviderImpl<? extends T> internalProvider, boolean isBound) {
+  private <T> InternalProviderImpl installInternalProvider(Class<T> clazz, String bindingName, InternalProviderImpl<? extends T> internalProvider,
+      boolean isBound) {
     Map<Class, UnNamedAndNamedProviders> map;
     if (isBound) {
       map = mapClassesToAllProviders;
     } else {
       map = mapClassesToUnBoundProviders;
     }
 
-    synchronized (clazz) {
-      UnNamedAndNamedProviders<T> unNamedAndNamedProviders = map.get(clazz);
-      if (unNamedAndNamedProviders == null) {
-        unNamedAndNamedProviders = new UnNamedAndNamedProviders<>();
-        map.put(clazz, unNamedAndNamedProviders);
-      }
-      if (bindingName == null) {
-        unNamedAndNamedProviders.setUnNamedProvider(internalProvider);
-      } else {
-        Map<String, InternalProviderImpl<? extends T>> mapNameToProvider = unNamedAndNamedProviders.getMapNameToProvider();
-        if (mapNameToProvider == null) {
-          mapNameToProvider = new HashMap<>();
-          unNamedAndNamedProviders.setMapNameToProvider(mapNameToProvider);
+    UnNamedAndNamedProviders<T> unNamedAndNamedProviders = map.get(clazz);
+    if (unNamedAndNamedProviders == null) {
+      synchronized (map) {
+        unNamedAndNamedProviders = map.get(clazz);
+        if (unNamedAndNamedProviders == null) {
+          unNamedAndNamedProviders = new UnNamedAndNamedProviders<>();
+          mapClassesToAllProviders.put(clazz, unNamedAndNamedProviders);
```

Comment:
you may need to synchronized `mapClassesToAllProviders`. the `map` here may only synchronized `mapClassesToUnBoundProviders` not `mapClassesToAllProviders`